desc/integrals/bounce_integral.py

```diff
@@ -1,26 +1,740 @@
+        Shape (..., grid.num_theta // 2 + 1, grid.num_zeta)
+        Complex coefficients of 2D real FFT of ``f``.
+
+    """
+    if not is_reshaped:
+        f = grid.meshgrid_reshape(f, "rtz")
+    # real fft over poloidal since usually m > n
+    return rfft2(f, axes=(-1, -2), norm="forward")
+
+
+def _transform_to_clebsch(grid, nodes, f, is_reshaped=False):
+    """Transform to Clebsch spectral domain.
+
+    Parameters
+    ----------
+    grid : Grid
+        Tensor-product grid in (θ, ζ) with uniformly spaced nodes in
+        (2π × 2π) poloidal and toroidal coordinates.
+        Preferably power of 2 for ``grid.num_theta`` and ``grid.num_zeta``.
+    nodes : jnp.ndarray
+        Shape (L, M, N, 2) or (M, N, 2).
+        DESC coordinates (θ, ζ) sourced from the Clebsch coordinates
+        ``FourierChebyshevBasis.nodes(M,N,domain=(0,2*jnp.pi))``.
+    f : jnp.ndarray
+        Function evaluated on ``grid``.
+
+    Returns
+    -------
+    a : FourierChebyshevSeries
+        Spectral coefficients of f(α, ζ).
+
+    """
+    assert nodes.shape[-1] == 2
+    if not is_reshaped:
+        f = grid.meshgrid_reshape(f, "rtz")
+
+    M, N = nodes.shape[-3], nodes.shape[-2]
+    return FourierChebyshevSeries(
+        f=interp_rfft2(
+            # Interpolate to nodes in Clebsch space,
+            # which is not a tensor product node set in DESC space.
+            xq=nodes.reshape(*nodes.shape[:-3], M * N, 2),
+            f=f[..., jnp.newaxis, :, :],
+            axes=(-1, -2),
+        ).reshape(*nodes.shape[:-3], M, N),
+        domain=(0, 2 * jnp.pi),
+    )
+
+
+def _transform_to_clebsch_1d(grid, alpha, theta, B, N_B, is_reshaped=False):
+    """Transform to single variable Clebsch spectral domain.
+
+    Parameters
+    ----------
+    grid : Grid
+        Tensor-product grid in (θ, ζ) with uniformly spaced nodes in
+        (2π × 2π) poloidal and toroidal coordinates.
+        Preferably power of 2 for ``grid.num_theta`` and ``grid.num_zeta``.
+    alpha : jnp.ndarray
+        Shape (L, num_transit) or (num_transit, ).
+        Sequence of poloidal coordinates A = (α₀, α₁, …, αₘ₋₁) that specify field line.
+    theta : jnp.ndarray
+        Shape (L, M, N) or (M, N).
+        DESC coordinates θ sourced from the Clebsch coordinates
+        ``FourierChebyshevBasis.nodes(M,N,domain=(0,2*jnp.pi))``.
+    B : jnp.ndarray
+        |B| evaluated on ``grid``.
+    N_B : int
+        Desired Chebyshev spectral resolution for |B|. Preferably power of 2.
+
+    Returns
+    -------
+    T, B : tuple[PiecewiseChebyshevSeries]
+        Set of 1D Chebyshev spectral coefficients of θ along field line.
+        {θ_α : ζ ↦ θ(α, ζ) | α ∈ A} where A = (α₀, α₁, …, αₘ₋₁) is ``alpha``.
+        Likewise with |B|.
+
+    """
+    if not is_reshaped:
+        B = grid.meshgrid_reshape(B, "rtz")
+
+    # Evaluating set of single variable maps is more efficient than evaluating
+    # multivariable map, so we project θ to a set of Chebyshev series.
```

Comment:
If `num_transit` > `M//2 + 1` then this increases the size of `T`, but it allows computing $\theta(\alpha, \zeta)$ at a given quadrature point with `N` multiplication and sums rather than `(M//2 + 1) * N`. Given that `B` already has size ``num_transit * N_B``, I figured the memory for `T` is not limiting.